### PR TITLE
Add publish_not_ready_addresses to kubernetes_service

### LINKED ISSUE #305: Missing attribute publishNotReadyAddresses in kubernetes_service resource
_This issue was originally opened by @manthila as hashicorp/terraform#20256. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi All

We are using below terraform version.

### Current Terraform Version

```
Terraform v0.11.11
+ provider.kubernetes v1.5.0
```

### Use-cases
When using headless services for statefulsets, we set **publishNotReadyAddresses** to **true** in YAML syntax. The primary use case for setting this field is to use a StatefulSet’s Headless Service to propagate SRV records for its Pods without respect to their readiness for purpose of peer discovery.
We are re-writing scripts using terraform and found that this configuration is not available on terraform. Therefore the application is failing to start.

### Attempted Solutions

Therefor we are requesting for this attribute be presented on terraform.

Thank you

### PR DESCRIPTION
**Pull request by pdecat**
Resolves #305 

Acceptance tests results on minikube:

```
# make testacc TEST=./kubernetes TESTARGS='-run=TestAccKubernetesService_.* -count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./kubernetes -v -run=TestAccKubernetesService_.* -count=1 -timeout 120m
=== RUN   TestAccKubernetesService_basic
--- PASS: TestAccKubernetesService_basic (0.26s)
=== RUN   TestAccKubernetesService_loadBalancer
--- SKIP: TestAccKubernetesService_loadBalancer (0.01s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_nodePort
--- PASS: TestAccKubernetesService_nodePort (0.15s)
=== RUN   TestAccKubernetesService_noTargetPort
--- SKIP: TestAccKubernetesService_noTargetPort (0.01s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_stringTargetPort
--- SKIP: TestAccKubernetesService_stringTargetPort (0.00s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_externalName
--- PASS: TestAccKubernetesService_externalName (0.11s)
=== RUN   TestAccKubernetesService_importBasic
--- PASS: TestAccKubernetesService_importBasic (0.13s)
=== RUN   TestAccKubernetesService_generatedName
--- PASS: TestAccKubernetesService_generatedName (0.13s)
=== RUN   TestAccKubernetesService_importGeneratedName
--- PASS: TestAccKubernetesService_importGeneratedName (0.14s)
PASS
ok      github.com/terraform-providers/terraform-provider-kubernetes/kubernetes 0.976s
```

### CODE REVIEW

File: kubernetes/structure_service_spec.go [hunk 1]
```diff
@@ -54,6 +54,10 @@ func flattenServiceSpec(in v1.ServiceSpec) []interface{} {
 	if in.ExternalName != "" {
 		att["external_name"] = in.ExternalName
 	}
+	if in.PublishNotReadyAddresses != false {
```
Comment (alexsomesan) [inline kubernetes/structure_service_spec.go]: Why do we need this check?

Comment (pdecat) [reply kubernetes/structure_service_spec.go]: I was asking myself the same question then I took example from https://github.com/terraform-providers/terraform-provider-kubernetes/blob/master/kubernetes/structure_persistent_volume_spec.go#L20

**Comment by pdecat** [comment]
Acceptance tests do not pass on GKE 1.11.6:

```
=== RUN   TestAccKubernetesService_basic
--- FAIL: TestAccKubernetesService_basic (2.39s)
    testing.go:538: Step 1 error: Error applying: 1 error occurred:
                * kubernetes_service.test: 1 error occurred:
                * kubernetes_service.test: Failed to update service: jsonpatch replace operation does not apply: doc is missing key: /spec/publishNotReadyAddresses
```

File: kubernetes/structure_service_spec.go (hunk 1 above)
Comment (pdecat) [reply kubernetes/structure_service_spec.go]: Removed the check.

**Comment by pdecat** [comment]
Acceptance tests results on minikube still pass with last fix:

```
# make testacc TEST=./kubernetes TESTARGS='-run=TestAccKubernetesService_.* -count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./kubernetes -v -run=TestAccKubernetesService_.* -count=1 -timeout 120m
=== RUN   TestAccKubernetesService_basic
--- PASS: TestAccKubernetesService_basic (0.25s)
=== RUN   TestAccKubernetesService_loadBalancer
--- SKIP: TestAccKubernetesService_loadBalancer (0.01s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_nodePort
--- PASS: TestAccKubernetesService_nodePort (0.14s)
=== RUN   TestAccKubernetesService_noTargetPort
--- SKIP: TestAccKubernetesService_noTargetPort (0.01s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_stringTargetPort
--- SKIP: TestAccKubernetesService_stringTargetPort (0.01s)
    provider_test.go:207: The Kubernetes endpoint must come from an environment which supports load balancer provisioning for this test to run - skipping
=== RUN   TestAccKubernetesService_externalName
--- PASS: TestAccKubernetesService_externalName (0.11s)
=== RUN   TestAccKubernetesService_importBasic
--- PASS: TestAccKubernetesService_importBasic (0.16s)
=== RUN   TestAccKubernetesService_generatedName
--- PASS: TestAccKubernetesService_generatedName (0.13s)
=== RUN   TestAccKubernetesService_importGeneratedName
--- PASS: TestAccKubernetesService_importGeneratedName (0.16s)
PASS
ok      github.com/terraform-providers/terraform-provider-kubernetes/kubernetes 1.000s
```

**Comment by pdecat** [comment]
Acceptance tests results on GKE now pass with last fix:

```
# make testacc TEST=./kubernetes TESTARGS='-run=TestAccKubernetesService_.* -count=1'                                                                                                  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./kubernetes -v -run=TestAccKubernetesService_.* -count=1 -timeout 120m
=== RUN   TestAccKubernetesService_basic
--- PASS: TestAccKubernetesService_basic (2.53s)
=== RUN   TestAccKubernetesService_loadBalancer
--- PASS: TestAccKubernetesService_loadBalancer (49.72s)
=== RUN   TestAccKubernetesService_nodePort
--- PASS: TestAccKubernetesService_nodePort (1.43s)

```

File: kubernetes/structure_service_spec.go [hunk 2]
```diff
@@ -207,10 +205,18 @@ func patchServiceSpec(keyPrefix, pathPrefix string, d *schema.ResourceData, v *v
 		})
 	}
 	if d.HasChange(keyPrefix + "publish_not_ready_addresses") {
-		ops = append(ops, &ReplaceOperation{
-			Path:  pathPrefix + "publishNotReadyAddresses",
-			Value: d.Get(keyPrefix + "publish_not_ready_addresses").(bool),
-		})
+		p := pathPrefix + "publishNotReadyAddresses"
+		v := d.Get(keyPrefix + "publish_not_ready_addresses").(bool)
+		if v {
```
Comment (alexsomesan) [inline kubernetes/structure_service_spec.go]: Fancy! :)

**Comment by pdecat** [comment]
Looks like I broke the corresponding datasource with this change:

```
=== RUN   TestAccKubernetesDataSourceService_basic
--- FAIL: TestAccKubernetesDataSourceService_basic (0.09s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
                * data.kubernetes_service.test: data.kubernetes_service.test: Invalid address to set: []string{"spec", "0", "publish_not_ready_addresses"}


    testing.go:599: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.

        Error: Error refreshing: 1 error occurred:
                * data.kubernetes_service.test: 1 error occurred:
                * data.kubernetes_service.test: data.kubernetes_service.test: Invalid address to set: []string{"spec", "0", "publish_not_ready_addresses"}





        State: <nil>
```

**Pull request merged**